core/src/main/java/com/linecorp/armeria/client/ClientRequestContextBuilder.java

```diff
@@ -103,29 +116,35 @@ public ClientRequestContextBuilder requestOptions(RequestOptions requestOptions)
      * Returns a new {@link ClientRequestContext} created with the properties of this builder.
      */
     public ClientRequestContext build() {
-        final Endpoint endpoint;
-        if (this.endpoint != null) {
-            endpoint = this.endpoint;
+        final EndpointGroup endpointGroup;
+        if (this.endpointGroup != null) {
+            endpointGroup = this.endpointGroup;
         } else {
-            endpoint = Endpoint.parse(authority());
+            endpointGroup = Endpoint.parse(authority());
         }
 
         final CancellationScheduler responseCancellationScheduler;
         if (timedOut()) {
             responseCancellationScheduler = CancellationScheduler.finished(false);
         } else {
             responseCancellationScheduler = CancellationScheduler.ofClient(0);
-            responseCancellationScheduler.initAndStart(eventLoop(), noopCancellationTask);
         }
-
         final DefaultClientRequestContext ctx = new DefaultClientRequestContext(
-                eventLoop(), meterRegistry(), sessionProtocol(),
-                id(), method(), requestTarget(), options, request(), rpcRequest(),
-                requestOptions, responseCancellationScheduler,
+                eventLoop(), meterRegistry(), sessionProtocol(), id(), method(), requestTarget(), options,
+                request(), rpcRequest(), requestOptions, responseCancellationScheduler,
                 isRequestStartTimeSet() ? requestStartTimeNanos() : System.nanoTime(),
                 isRequestStartTimeSet() ? requestStartTimeMicros() : SystemInfo.currentTimeMicros());
+        if (!timedOut()) {
+            ctx.whenInitialized().handle((unused1, unused2) -> {
+                ctx.responseCancellationScheduler().initAndStart(ctx.eventLoop(), noopCancellationTask);
+                return null;
+            });
+        }
 
-        ctx.init(endpoint);
+        ctx.init(endpointGroup).handle((unused, cause) -> {
+            ctx.finishInitialization(cause == null);
+            return null;
+        });
```

Comment:
I think we can just do
```java
ctx.init(endpointGroup).handle((unused, cause) -> {
    ctx.finishInitialization(cause == null);
    if (!timedOut()) {
        ctx.responseCancellationScheduler().initAndStart(ctx.eventLoop(), noopCancellationTask);
    }
    return null;
});
```

Reply:
It's much more readable and concise.
Why didn't I think of that? It's really amazing!🥳